### PR TITLE
gamnit: simple 2D API, input events and the game Asteronits

### PR DESCRIPTION
**Pull request by xymus**
You should probably read this PR commit by commit, it is a bit long but it's mostly due to the sample game, a big XML file and changes to indentation.

The new API provided by `simple_2d` is meant to replace mnit for writing 2D games.
Client modules only have to create instances of `Sprite` which will be displayed as long as they are in `App::sprites`.

At this point, the `simple_2d` API is not optimized and you may see a slowdown at the insertion of new sprites. However, we could improve the performance by preallocating buffer space for more sprites, and by caching sprites state on the GPU to avoid unnecessary updates.

This PR also adds support for input events for Linux and Android. It uses the existing services from `mnit::input`, `sdl` and `android::input_events`.

Finally, the game Asteronits is an example for the `simple_2d` API. In this game, the player controls a ship, avoids asteroids and destroys them. The gameplay and UI is limited by design to be a simple example. Note that the `touch_ui` variant is used on Android only but it also works on Linux.

The tool texture_atlas_parser reads XML TextureAtlas files and produces Nit classes with one attribute per subtexture. We could move this tool if it is useful for any other projects.

### CODE REVIEW

File: lib/mnit/input.nit [hunk 1]
```diff
@@ -34,7 +35,12 @@ interface PointerEvent
 
 	# Is down? either going down or already down
 	fun pressed: Bool is abstract
+
+	# Is it not currently pressed down?
 	fun depressed: Bool is abstract
```
Comment (privat) [inline lib/mnit/input.nit]: you should document the relation with `pressed`; by the way, why the default implementation is not `return not pressed`?

File: lib/gamnit/textures.nit [hunk 2]
```diff
@@ -167,16 +153,32 @@ class GamnitSubtexture
...
 	redef fun load(force) do root.load(force)
 
-	redef var offset_left = 0.0
-	redef var offset_top = 0.0
-	redef var offset_right = 1.0
-	redef var offset_bottom = 1.0
+	redef var offset_left = parent.offset_left + left / root.width is lazy
+	redef var offset_top = parent.offset_top + top / root.height is lazy
+	redef var offset_right = offset_left + width / root.width is lazy
+	redef var offset_bottom = offset_top + height / root.height is lazy
```
Comment (privat) [inline lib/gamnit/textures.nit]: why lazy?

Comment (xymus) [inline lib/gamnit/textures.nit]: Dimensions of root textures are known only when they are loaded.

File: lib/gamnit/simple_2d.nit [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+
+	# Scaling per vertex
+	var scale = attributes["scale"].as(AttributeFloat) is lazy
+
+	# Model view projection matrix
+	var mvp = uniforms["mvp"].as(UniformMat4) is lazy
+end
+
+redef class Point3d[N]
+	fun offset(x, y, z: Numeric): Point3d[Float]
```
Comment (privat) [inline lib/gamnit/simple_2d.nit]: move up to geometry?

File: lib/gamnit/gamnit_linux.nit [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+import display_linux
+
+redef class App
+	redef fun feed_events
+	do
+		var display = display
+		if display == null then return
+
+		var events = display.sdl_display.events
+		display.check_lock_cursor # HACK
```
Comment (privat) [inline lib/gamnit/gamnit_linux.nit]: I don't understand the hack

File: lib/gamnit/examples/asteronits/Makefile [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+bin/texture_atlas_parser: src/texture_atlas_parser.nit
+	${NITC} src/texture_atlas_parser.nit -o $@
+
+src/controls.nit: art/controls.svg
+	../../../../contrib/inkscape_tools/bin/svg_to_png_and_nit art/controls.svg -a assets/ -s src/ -x 2.0 -g
+
+src/spritesheet_city.nit: bin/texture_atlas_parser
+	bin/texture_atlas_parser art/sheet.xml --dir src/ -n spritesheet
+
+prepare: src/controls.nit src/spritesheet_city.nit
```
Comment (privat) [inline lib/gamnit/examples/asteronits/Makefile]: by `prepare`, do you mean `pre-build`? cf #1749

File: lib/gamnit/examples/asteronits/README.md [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+
+* `game_logic` defines the pure game logic without any of the display details.
+
+* `asteronits` implements the display logic with support for keyboard input events.
+
+* `touch_ui` adds an interface for touchscreen devices with buttons to open fire and control the ship.
+
+# Art
+
+Artwork credit: Kenney.nl
```
Comment (privat) [inline lib/gamnit/examples/asteronits/README.md]: website? license?

File: lib/gamnit/simple_2d.nit (hunk 3 above)
Comment (xymus) [inline lib/gamnit/simple_2d.nit]: Because it's Float specific I prefer to leave it here. But we can move and update it when needed.

**Comment by xymus** [comment]
Rerolled with comments and to fix tests.

**Comment by xymus** [comment]
Note that the Asteronits example is functional but incomplete. I plan to add particles (with the corresponding gamnit service), audio (when app::audio is decoupled from mnit), a static background with perspective and the globe example as background.

**Comment by xymus** [comment]
Rerolled to fix doc generated by texture_atlas_parser.

**Comment by privat** [comment]
I think you should move the game to contrib instead of examples

**Comment by privat** [comment]
the controls on android are just awful. +1

File: lib/gamnit/examples/asteronits/src/asteronits.nit [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Portable game to destroy asteroids
+module asteronits is
+	app_name "Asteronits"
+	app_namespace "org.nitlanguage.asteronits"
+	app_version(1, 0, git_revision)
+
+	android_manifest_activity """android:screenOrientation="sensorLandscape""""
```
Comment (R4PaSs) [inline lib/gamnit/examples/asteronits/src/asteronits.nit]: I really need to find a way to properly parse this.

**Comment by xymus** [comment]
So make checks do not declare "NIT_TESTING"? Or should I not execute the game in the check rule?

**Comment by privat** [comment]
`make check` should be executable in a headless environment without any prior configuration, so that a human or integration server can just execute it.

**Comment by privat** [comment]
some check failed. http://gresil.org/jenkins/job/nit_test_tools/930/testReport/

**Comment by xymus** [comment]
Errors on mongodb connections? or is it a wrong link?

**Comment by xymus** [comment]
The last commit should fix the errors from http://www.google.com/url?q=http%3A%2F%2Fgresil.org%2Fjenkins%2Fjob%2Fnit_test_android%2F53%2F&sa=D&sntz=1&usg=AFQjCNH64gMliHWvQ_GuLstD3-mJPSaE_Q

**Comment by privat** [comment]
still broken http://gresil.org/jenkins/job/nit_test_android/56/testReport/

**Comment by xymus** [comment]
These Makefiles are getting complex, but it should be fixed now.

**Pull request merged**